PyRBP.Features
==================================


.. py:method:: PyRBP.Features.generateBPFeatures(sequences, pseudoKNC=False, ktuple=3, zigzag_coding=False, guanine_cytosine_Quantity=False, nucleotide_tilt=False, percentage_of_bases=False, PGKM=False, gapValue=1, kValue=2, mValue=2, DPCP=False)

    This function is used to generate various types of features based on the sequences (physicochemical features, sequence properties, base composition, etc.).

    :Parameters:
                .. class:: sequences:list or array, necessary parameters

                        List or array of sequences used to generate features.

                .. class:: pseudoKNC:bool, default=False

                        Whether to use pseudoKNC algorithm to generate features, if ``True``, the value of the parameter ktuple will be used for subsequent feature generation.

                .. class:: ktuple:int, default=3

                        Used to determine the number of bases for each tuple in `pseudoKNC` in the values of ``[3, 4, 5]``.

                .. class:: zigzag_coding:bool, default=False

                        Whether to use zigzag_coding algorithm to generate features.

                .. class:: guanine_cytosine_Quantity:bool, default=False

                        Whether to use guanine_cytosine_Quantity algorithm to generate features.

                .. class:: nucleotide_tilt:bool, default=False

                        Whether to use nucleotide_tilt algorithm to generate features.

                .. class:: percentage_of_bases:bool, default=False

                        Whether to use percentage_of_bases algorithm to generate features.

                .. class:: PGKM:bool, default=False

                        Whether to use PGKM algorithm to generate features. if ``True``, the value of the parameter gapValue, kValue and mValue will be used for subsequent feature generation.

                .. class:: gapValue:int, default=1

                        Used to determine the number of gaps between each two tuples. Its value is limited to ``[1, 2, 3, 4, 5]``.

                .. class:: kValue:int, default=2

                        Used to determine the number of bases for the first tuple in `PGKM`, the value is limited to ``[1, 2]``.

                .. class:: mValue:int, default=2

                        Used to determine the number of bases for the second tuple in `PGKM`, the value is limited to ``[1, 2]``.

                .. class:: DPCP:bool, default=False

                        Whether to use DPCP algorithm to generate features.
    :Attributes:
                .. class:: features:list of feature according to each sequence

                        Used to store the final generated features for each sequence. When the function returns, it has been converted to an array.

                .. class:: feature:list of feature values for one sequence

                        Used Used to store the various features of the sequence being processed.

    .. note:: The values of gapValue, kValue and mValue greatly affect the number of feature dimensions generated by the PGKM algorithm, the larger the values of these three parameters, the longer the running time of PGKM. Please set the parameter values according to your needs.



.. py:function:: PyRBP.Features.generateDynamicLMFeatures(sequences, kmer=3, model='')

    This function is used to generate the dynamic semantic information matrix of the sequence, in which we provide the fine-tuned ``BERT models (RBPBERT)`` for the RBP classification problem. The models can be downloaded from `figshare link <https://figshare.com/articles/software/LM_for_PyRBP/21383187>`_. When extraction is complete, the dimension of the feature matrix obtained is (``number of sequences``, ``number of tokens per sequence``, ``768``), where ``768`` indicates 768 hidden units of 12 attention heads in the last transformer layer.

    :Parameters:
                .. class:: sequences:list or array, necessary parameters

                        List or array of sequences used to generate features.

                .. class:: kmer:int, default=3

                        kmer specifies the window size to be used when tokenizing sequences. There are four window sizes to choose from ``[3, 4, 5, 6]``.

                .. class:: model:str, default=''

                        The path where the downloaded ``RBPBERT`` model is stored, it should be noted that when passing parameters, only the `absolute path` to the folder where the model is located should be passed in, as in the example:
                        ::

                            /home/wangyansong/PyRBP/src/dynamicRNALM/circleRNA/pytorch_model_3mer


.. py:function:: PyRBP.Features.generateStaticLMFeatures(sequences, kmer=3, model='')

    There are four static semantic models available in the PyRBP: ``fasttext``, ``GloVe``, ``word2vec`` and ``doc2vec``. These models can be downloaded from `figshare link <https://figshare.com/articles/software/LM_for_RBP_package/21383187>`_.

    :Parameters:
                .. class:: sequences:list or array, necessary parameters

                        List or array of sequences used to generate features.

                .. class:: kmer:int, default=3

                        kmer specifies the window size to be used when tokenizing sequences. There are four window sizes to choose from ``[3, 4, 5, 6]``.

                .. class:: model:str, default=''

                        The path where the downloaded ``static semantic model`` is stored, it should be noted that when passing parameters, you need to be careful that the model path passed in is consistent with ``kmer``.

    :Attributes:
                .. class:: LM_type:str, ['word2vec', 'fasttext', 'doc2vec', 'GloVe']

                        Separated from the incoming model file name, used to distinguish different models when extracting embedding.

    .. note:: Since both of the above modules need to separate some required information from the model names to perform matching checks, please do not make any changes to the downloaded model file names when using dynamic or static language models for semantic feature extraction, otherwise it may cause the module not to work properly.



.. py:function:: PyRBP.Features.generateStructureFeatures(dataset_path='', script_path='', basic_path='',  W=101, L=70, u=1, dataset_name='')

    This function calls ``RNAplfold`` to calculate locally stable secondary structure - pair probabilities [RNAplfold]_. The respective scripts of ``RNAplfold`` can be downloaded at `figshare link <https://figshare.com/articles/software/LM_for_RBP_package/21383187>`_.

    :Parameters:
                .. class:: dataset_path:str, default=''

                        Path to the fasta file.

                .. class:: script_path:str, default=''

                        The path where the ``RNAplfold`` scripts are located.

                .. class:: basic_path:str, default=''

                        The path where the generated secondary structure profile files are stored, under which four folders E, H, I and M will be created, as well as the final structure information file combined_profile.txt

                .. class:: W:int, default=101

                        Average the pair probabilities over windows of given size.

                .. class:: L:int, default=70

                        Set the maximum allowed separation of a base pair to span.
                        By setting the maximum base pair span no pairs (i,j) with jâˆ’i > span will be allowed. Defaults to winsize if parameter is omitted.

                .. class:: u:int, default=1

                        Compute the mean probability that regions of length 1 to a given length are unpaired

                .. class:: dateset_name, default=''

                        To facilitate the storage of structural information for multiple datasets, dataset_name is used to mark different datasets, which is combined with ``basic_path`` to form the storage path.

    :Attributes:
                .. class:: path:str

                        Used to store the final secondary structure information file.

                .. class:: E_path:str

                        Used to store the E_RNAplfold secondary structure information file.

                .. class:: M_path:str

                        Used to store the M_RNAplfold secondary structure information file.

                .. class:: I_path:str

                        Used to store the I_RNAplfold secondary structure information file.

                .. class:: H_path:str

                        Used to store the H_RNAplfold secondary structure information file.

                .. class:: cmd:str

                        Used to store the command to summary secondary structure information file.

    .. note:: Note that you need to give the RNAplfold scripts executable permissions using the following command:
                ::

                    chmod 764 path_to_the_scripts



    .. [RNAplfold] R. Lorenz, S.H. Bernhart, C. Hoener zu Siederdissen, H. Tafer, C. Flamm, P.F. Stadler and I.L. Hofacker (2011), "ViennaRNA Package 2.0", Algorithms for Molecular Biology: 6:26